### PR TITLE
Issue #743: Correctly handle escaped slashes in JSON byte data

### PR DESCRIPTION
**Pull request by tbkka**
Protobuf JSON is designed to be easily implemented on top of standard
JSON encoders.  For historical reasons, such encoders often encode "/"
(which commonly occurs in base-64 data) as "\/".

I overlooked support for backslash escapes when parsing byte data.

This adds tests and support for all the simple backslash escapes.
It turns out that "\/" is the only one that is legal in base-64 data,
so we simply fail if we encounter \b \f \n \r \t \" \\ or any invalid
sequence.

Note: This does not support \u1234 hexadecimal escapes.  (I don't
expect to ever see those in practice, so I doubt this is a problem.)

### CODE REVIEW

File: Sources/SwiftProtobuf/JSONScanner.swift [hunk 1]
```diff
@@ -172,26 +190,24 @@ private func parseBytes(
...
                 case asciiDoubleQuote:
                     source.formIndex(after: &index)
                     break digits
+                case asciiBackslash:
+                    source.formIndex(after: &index)
+                    let escaped = source[index]
+                    switch escaped {
+                    case asciiForwardSlash:
+                        k = base64Values[Int(escaped)]
+                    default:
```
Comment (thomasvl) [inline Sources/SwiftProtobuf/JSONScanner.swift]: This can't actually happen since the loop before should have blocked it?

File: Sources/SwiftProtobuf/JSONScanner.swift [hunk 2]
```diff
@@ -134,14 +134,32 @@ private func parseBytes(
...
+            if index == end {
+                throw JSONDecodingError.malformedString
+            }
+            let escaped = source[index]
+            switch escaped {
+            case asciiLowerU:
+                // TODO: Caller should pre-parse with full string decoder and then try again
+                throw JSONDecodingError.malformedString
+            case asciiForwardSlash:
+                rawChars += 1
```
Comment (cburrows) [inline Sources/SwiftProtobuf/JSONScanner.swift]: Should you set `sawSection4Characters = true` here? Why not just reassign digit and fall through to the else?

File: Sources/SwiftProtobuf/JSONScanner.swift [hunk 3]
```diff
@@ -172,26 +190,24 @@ private func parseBytes(
...
+                    let escaped = source[index]
+                    switch escaped {
+                    case asciiForwardSlash:
+                        k = base64Values[Int(escaped)]
+                    default:
+                        throw JSONDecodingError.malformedString
+                    }
                 case asciiSpace:
-                    break
+                    source.formIndex(after: &index)
```
Comment (cburrows) [inline Sources/SwiftProtobuf/JSONScanner.swift]: Why this change? It appears to make spaces permissible, and ignored, at most locations in the middle of the base64 data?

File: Sources/SwiftProtobuf/JSONScanner.swift [hunk 4]
```diff
@@ -172,26 +190,24 @@ private func parseBytes(
...
-                    n = 0
-                }
-            } else {
+            var k = base64Values[Int(digit)]
+            if k < 0 {
                 switch digit {
                 case asciiDoubleQuote:
                     source.formIndex(after: &index)
                     break digits
+                case asciiBackslash:
```
Comment (cburrows) [inline Sources/SwiftProtobuf/JSONScanner.swift]: While we are here, I notice that the protobuf documentation permits the base64 string to be URL-safe, i.e. `+` and `=` and `/` can by `%2B` or whatever. Maybe open another issue?

File: Sources/SwiftProtobuf/JSONScanner.swift (hunk 3 above)
Comment (thomasvl) [reply Sources/SwiftProtobuf/JSONScanner.swift]: It is because of the reflow of the loop, this now does a the `continue digits` where before it just did a `break` and so it fell to the code below that did this work.

File: Sources/SwiftProtobuf/JSONScanner.swift (hunk 4 above)
Comment (thomasvl) [reply Sources/SwiftProtobuf/JSONScanner.swift]: They are both mixed in the table, the section 4/5 checks are to ensure we only ever get one set and not both (if I'm remembering correctly)

File: Sources/SwiftProtobuf/JSONScanner.swift (hunk 2 above)
Comment (thomasvl) [reply Sources/SwiftProtobuf/JSONScanner.swift]: Oops, yup, seems like it should have been done.

File: Sources/SwiftProtobuf/JSONScanner.swift (hunk 4 above)
Comment (tbkka) [reply Sources/SwiftProtobuf/JSONScanner.swift]: "URL-safe base-64" is not the same as "URL encoding."  The URL-safe alphabet is the one that uses "-" and "_" instead of "+" and "/".  (This is formalized in RFC4648 section 5 as the "Base 64 Encoding with URL and Filename Safe Alphabet" aka "base64url".)

I've expanded the comments above the table to explain this further.

File: Sources/SwiftProtobuf/JSONScanner.swift (hunk 3 above)
Comment (tbkka) [reply Sources/SwiftProtobuf/JSONScanner.swift]: This is based on the Google C++ implementation, which is tolerant of spaces.  On the other hand, [RFC4648 Section 3.3](https://tools.ietf.org/html/rfc4648#section-3.3) urges people to reject base64 text with spaces.  I've chosen to ignore spaces to better match Google's implementation, but this would be easy to change if we had a reason.

File: Sources/SwiftProtobuf/JSONScanner.swift (hunk 4 above)
Comment (cburrows) [reply Sources/SwiftProtobuf/JSONScanner.swift]: Ah, my mistake. Thanks.

File: Sources/SwiftProtobuf/JSONScanner.swift (hunk 3 above)
Comment (tbkka) [reply Sources/SwiftProtobuf/JSONScanner.swift]: I added a test to document this behavior.

File: Sources/SwiftProtobuf/JSONScanner.swift (hunk 2 above)
Comment (tbkka) [reply Sources/SwiftProtobuf/JSONScanner.swift]: Good catch!  I've fixed this and added a test for it.
